### ISSUE TITLE
[clang enhancement] improve -Wreturn-type-c-linkage warning message

### ISSUE THREAD

**Issue by illusory0x0** [llvm/llvm-project#137282]
When I write the code below, clang report warning, but `struct Foo` is complete type. the Foo<int> doesn't instantiate template. We can instantiate this type in C++ linkage scope, then Most of complier can complies successfully.

### Expected Behavior

add more usefull infomation help user to solve.

```
try to instantiate template in C++ linkage, if this excectly not incomplete type.
```

### Actual Behavior

```
<source>:14:14: warning: 'foo' has C-linkage specified, but returns incomplete type 'Foo<int>' which could be incompatible with C [-Wreturn-type-c-linkage]
```

```C++
template<typename T>
struct Foo {
    T* ptr;
};


void force_instantiate_template() {
    // Foo<int> x;
    // If uncooment the line above, clang woundn't report warning.
    // and MSVC will complie successfully.
}

extern "C" {
    Foo<int> foo() {
        return { .ptr = nullptr } ;
    }

    void bar(Foo<int>) {

    }
}
```

[godbolt](https://godbolt.org/z/e54Gb61Kx)

**Comment by frederick-vs-ja** [comment]
Isn't this just a false positive?

**Comment by illusory0x0** [comment]
> Isn't this just a false positive?

I'm not sure what is the correct behavior, But MSVC always complie error [C2526](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-errors-2/compiler-error-c2526?view=msvc-170). Maybe we also report issues in MSVC?
